### PR TITLE
[crypto] Set bits of candidate primes in RSA keygen.

### PR DESCRIPTION
**Pull request by jadephilipoom**
FIPS 186-5 gives implementers a few new degrees of freedom when generating prime candidates in RSA key generation:
1. you can restrict prime candidates to any specific odd value mod 8
2. you can set the highest 2 bits arbitrarily

Using (1) to pick prime candidates that are 3 mod 4 makes constant-time Miller-Rabin a lot simplier; in general form, you have to check if the remainder of the bits of `p - 1` are 0, and do different things depending on which case you're in. But if the prime candidate is 3 mod 4, then `p-1` always has a 1 bit in it until the very last bit, and you can remove all that case analysis.

Using (2) to set the highest bits both to 1 allows one to ensure that candidate primes are always in range; one of the first checks a candidate prime needs to pass checks that the value is `>= sqrt(2) * 2^(nlen/2 - 1)`, where `nlen` is the bit-length of the RSA modulus. Since `sqrt(2) < 1.5`, setting the top two bits means the candidate prime always passes this check, which saves time and entropy by discarding fewer values and not having to do the range-check. It also means we don't need to store a large constant relating to the square root of 2 to calculate the bound precisely.

These changes make RSA key generation significantly faster and smaller:
- 9% reduction in binary size
  - `run_rsa_keygen.elf` has 3884 bytes of IMEM and 2080 bytes of DMEM on master, and after the change it has 3632 bytes of IMEM and 1824 bytes of DMEM
- 16% speedup in primality testing
  - primality testing takes by far the majority of time in RSA key generation
  - speedup estimate from `//sw/otbn/crypto/tests:primality_test`, which runs the primality test against a true prime;   1940418  cycles on master and 1670938 cycles after the change

In my opinion, the binary size change is the biggest win, because RSA key generation is quite close to the IMEM size limit and I now have the space to implement much more impactful speed optimizations (such as checking if the candidate prime is a multiple of small primes, which detects the vast majority of composites without running a primality test).

### CODE REVIEW

File: sw/otbn/crypto/primality.s [hunk 1]
```diff
@@ -198,51 +198,22 @@ miller_rabin_round:
...
- *
- *     // Get the lower bits (to see if they're all zero).
- *     w_low = (w - 1) mod (2^i)
- *
- *     // Determine if b is a witness to the primality of w.
- *     possibly_prime |= ((w_low == 0) && (z == w-1))
- *     possibly_prime |= ((w_low == 0) && (wi == 1) && (z == 1))
- *   }
- *   return possibly_prime
+ * If we specialize to the case that w mod 4 = 3, the routine becomes much
```
Comment (vrozic) [inline sw/otbn/crypto/primality.s:201]: Would it make sense to add a check that `w mod 4` is indeed 3.

**Comment by vrozic** [comment]
This is an excellent optimisation.
My only concern is that perhaps it creates a risk of fault-attacks. Something like:

1. Witness is set to w mod 8 = 3, `w = ...1000000011`
2. A fault is injected just before the primality test, setting `w = ...1000000001`
3. Because step 4.5 of the primality test is skipped, we end up with a composite number that passes primality test.

This can be mitigated with a redundant check, but perhaps there is already a high-level reason why we don't need to worry about this scenario.

**Comment by jadephilipoom** [comment]
> This can be mitigated with a redundant check, but perhaps there is already a high-level reason why we don't need to worry about this scenario.

I agree that for FI hardening it makes sense to have something ensure the candidate is 3 mod 4. Rather than a check, though, maybe we should just set those bits again in the primality test? That will consume less code size than a path that fails if they aren't set, and will still thwart the same class of attacks.

**Comment by vrozic** [comment]
> Rather than a check, though, maybe we should just set those bits again in the primality test? That will consume less code size than a path that fails if they aren't set, and will still thwart the same class of attacks.

Agreed. That is indeed a better solution.

**Review by vrozic — APPROVED** [review]

**Review by moidx — APPROVED** [review]

**Pull request merged**